### PR TITLE
[api-extractor] Fix incorrect indentation in emitted .d.ts/.api.md files

### PR DESCRIPTION
**Pull request by octogonz**
## Summary

Fixes https://github.com/microsoft/rushstack/issues/2792

## Details

Prior to this PR, the `Span` class was doing a lot juggling of whitespace tokens to ensure that the output is formatted correctly.  The difficult cases are (1) when API items get trimmed, ensuring that we don't delete the whitespace, and (2) when API items get sorted, ensuring that the newlines don't get sorted into the wrong places.

Consider this case from https://github.com/microsoft/rushstack/issues/2792:
```ts
export class ExampleA {
  private _member3: string;
  member2(): Promise<void>;
  member1: string;
}
```
The `_member3: string` has `<newline>` `<space>` `<space>` attached to it, whereas `member1: string;` has `<newline>` because it is followed by `}`.  In the API report, these get sorted alphabetically such that `_member3: string` ends up last.  The `Span.getModifiedText()` emitter carefully captures the `firstSeparator` and `lastSeparator` so that after sorting, `_member3` gets the `<newline>`:

https://github.com/microsoft/rushstack/blob/a57cb5634e95a1298a8e4b86a7db649d45925b76/apps/api-extractor/src/analyzer/Span.ts#L422-L426

The bug was that `_member3` is private and therefore does not get emitted in this case.  So we end up with `<newline>` `<space>` `<space>` `}` which is incorrect.

I tried to fix up those problems, but the implementation was just too complicated.  Instead this PR introduces a postprocessor called `IndentedWriter` that discards the original indentation and redoes the indentation. The logic is not too complicated because the compiler's .d.ts indentation is relatively simple and can be reproduced based on `{` `}` and `/*` `*/` pairs alone.  This assumes that the newlines are in the right places, so we cannot completely eliminate the `Span.getModifiedText()` heuristics.

I considered taking this a step further and invoking Prettier on the output, in which case we wouldn't need to do any formatting at all.  But that has some downsides:

- Prettier is a fairly big dependency and would impact API Extractor's runtime performance and also installation footprint
- Prettier would completely reformat the output, which might be undesirable for users who haven't adopted Prettier; currently API Extractor largely preserves the input formatting
- Eventually we'd like to implement source maps for .d.ts rollups (https://github.com/microsoft/rushstack/issues/1886); Prettier doesn't implement source maps, so it would make this feature much more difficult
- The .api.md syntax is not technically TypeScript; Prettier seems to handle it just fine, but it's not guaranteed

The PR happens to make it super easy for our `.d.ts` rollups to indent by 2 spaces instead of 4 spaces.  I almost included that change; however, I think we need to provide a setting for people who want 4 spaces, so that should be its own PR.

`IndentedWriter` is borrowed from [api-documenter/src/utils/IndentedWriter.ts](https://github.com/microsoft/rushstack/blob/master/apps/api-documenter/src/utils/IndentedWriter.ts) and probably should get pushed down into `node-core-library`, but I'd want to make it streamable before we do that. 

 

## How it was tested

I compared before/after for all .api.md files in this repo, and all .d.ts rollups from the build tests.

I compared before/after for all .api.md files and .d.ts rollups in our monorepo at work.

### CODE REVIEW

**Review by iclanton — APPROVED** [review]

File: apps/api-extractor/src/analyzer/Span.ts [hunk 1]
```diff
@@ -482,19 +489,49 @@ export class Span {
...
+
+    if (options.indentDocComment === IndentDocCommentState.AwaitingOpenDelimiter) {
+      let index: number = parsedText.indexOf('/*');
+      if (index >= 0) {
+        index += '/*'.length;
+        options.writer.write(parsedText.substring(0, index));
+        parsedText = parsedText.substring(index);
+        options.indentDocComment = IndentDocCommentState.AwaitingCloseDelimiter;
+
+        options.writer.increaseIndent(' ');
```
[the remark this answers is not in the document]
Comment (octogonz) [reply apps/api-extractor/src/analyzer/Span.ts:516]: It was originally designed that way, but this approach lets you do stuff like `writer.increaseIndent('// ')` or `writer.increaseIndent('> ')` which is a handy way to add a prefix to each line.

File: apps/api-extractor/src/generators/IndentedWriter.ts [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+ *     two
+ *   end
+ * ```
+ */
+export class IndentedWriter {
+  /**
+   * The text characters used to create one level of indentation.
+   * Two spaces by default.
+   */
+  public defaultIndentPrefix: string = '    ';
```
[the remark this answers is not in the document]
Comment (octogonz) [reply apps/api-extractor/src/generators/IndentedWriter.ts:38]: @iclanton  The compiler indents by 4 spaces, but I agree API Extractor should indent by 2 spaces.  However it's a somewhat big change, so we should provide a setting to restore the old behavior, which requires a separate PR.  I mentioned this in the PR description.

File: apps/api-extractor/src/analyzer/Span.ts [hunk 3]
```diff
@@ -2,8 +2,20 @@
 // See LICENSE in the project root for license information.
 
 import * as ts from 'typescript';
-import { StringBuilder } from '@microsoft/tsdoc';
-import { Sort } from '@rushstack/node-core-library';
+import { IndentedWriter } from '../generators/IndentedWriter';
+import { InternalError, Sort } from '@rushstack/node-core-library';
```
[the remark this answers is not in the document]
Comment (octogonz) [reply apps/api-extractor/src/analyzer/Span.ts:6]: Fixed

File: apps/api-extractor/src/analyzer/Span.ts [hunk 4]
```diff
@@ -407,7 +390,31 @@ export class Span {
...
+
+    if (this.modification.indentDocComment) {
+      if (options.indentDocComment === IndentDocCommentState.AwaitingCloseDelimiter) {
+        throw new InternalError('missing "*/" delimiter for comment block');
+      }
+      options.indentDocComment = IndentDocCommentState.Idle;
+    }
+  }
+
+  private _writeModifiedText2(options: IWriteModifiedTextOptions): void {
```
[the remark this answers is not in the document]
Comment (octogonz) [reply apps/api-extractor/src/analyzer/Span.ts:416]: It is a workaround for a `return` statement that was messing up the control flow. Lemme see if I can simplify it

[the remark this answers is not in the document]
Comment (octogonz) [reply apps/api-extractor/src/analyzer/Span.ts]: Fixed

**Pull request merged**